File: ProjectTwo(Python)/repositories/account_repo_impl.py
from exceptions.resource_not_found import ResourceNotFound
from models.accounts import Accounts
from repositories.account_repo import AccountRepo
from util.db_connection import connection


def _build_account(record):
    return Accounts(acct_id=record[0], user_id=record[1], account_type=record[2], account_balance=record[3])


class AcctRepoImpl(AccountRepo):
    def create_account(self, account):
        sql = 'INSERT INTO Accounts VALUES (DEFAULT, %s, %s, %s) RETURN *'
        cursor = connection.cursor()
        cursor.execute(sql, [account.acct_id, account.user_id, account.account_type, account.account_balance])
        connection.commit()
        record = cursor.fetchone()
        return _build_account(record)

    def get_account(self, acct_id):
        sql = 'SELECT * FROM Accounts WHERE acctID = %s'
        cursor = connection.cursor()
        cursor.execute(sql, [acct_id])
        record = cursor.fetchone()
        if record:
            return _build_account(record)
        else:
            raise ResourceNotFound(f"Account with ID {acct_id} - Not Found")

    def get_all_accounts(self, user_id):
        sql = "SELECT * from Accounts WHERE userID = %s"
        cursor = connection.cursor()
        cursor.execute(sql, [user_id])
        records = cursor.fetchall()
        if records:
            acct_list = [_build_account(record) for record in records]
            return acct_list
        else:
            raise ResourceNotFound(f"Account for ID {user_id} - Not Found")

    def update_account(self, change):
        sql = "UPDATE Accounts SET userID=%s, acctID=%s, accountType=%s, accountBalance=%s RETURN *"
        cursor = connection.cursor()
        cursor.execute(sql, [change.acct_id, change.user_id, change.account_type, change.account_balance])
        connection.commit()
        record = cursor.fetchone()
        return _build_account(record)

    def del_account(self, acct_id):
        sql = "DELETE FROM Accounts WHERE acctID=%s RETURN *"
        cursor = connection.cursor()
        cursor.execute(sql, [acct_id])
        connection.commit()
        return f"{acct_id} successfully deleted"


def _test():
    ar = AcctRepoImpl()
    print(ar.get_account(1))


if __name__ == '__main__':
    _test()

File: BankApp/src/main/java/repositories/AccountRepoImpl.java
package repositories;

import models.Account;
import util.JDBCConnection;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;


public class AccountRepoImpl implements AccountRepo {

    public static Connection conn = JDBCConnection.getConnection();

    @Override
    public Account getAccount(int id) {

        String sql = "SELECT * FROM accounts WHERE userId = ?";

        try {
            PreparedStatement ps = conn.prepareStatement(sql);
            ps.setInt(1, id);

            ResultSet rs = ps.executeQuery();

            if (rs.next()) {
                return buildAccount(rs);
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }

        return null;
    }

    @Override
    public List<Account> getAllAccounts() {

        String sql = "SELECT * FROM accounts";

        try {
            PreparedStatement ps = conn.prepareStatement(sql);

            ResultSet rs = ps.executeQuery();

            List<Account> accounts = new ArrayList<>();

            while(rs.next()) {
                accounts.add(buildAccount(rs));
            }

            return accounts;

        } catch (SQLException e) {
            e.printStackTrace();
        }

        return null;
    }

    @Override
    public Account addAccount(Account a) {

        String sql = "INSERT INTO accounts VALUES (DEFAULT,?,?,?) RETURNING *";

        try {
            PreparedStatement ps = conn.prepareStatement(sql);

            ps.setString(1, a.getAcct());
            ps.setDouble(2, a.getAcctBal());

            ResultSet rs = ps.executeQuery();

            if(rs.next()) {
                return buildAccount(rs);
            }


        } catch (SQLException e) {
            e.printStackTrace();
        }

        return null;
    }

    @Override
    public Account updateAccount(Account change) {
        try {
            String sql = "UPDATE accounts SET accounttype=?, accountBalance=?, WHERE acctid = ? RETURNING *";
            PreparedStatement ps = conn.prepareStatement(sql);

            ps.setString(1, change.getAcct());
            ps.setDouble(2, change.getAcctBal());
            ps.setInt(5, change.getUserId());

            ResultSet rs = ps.executeQuery();

            if (rs.next()) {
                return buildAccount(rs);
            }


        } catch (SQLException e) {
            e.printStackTrace();
        }

        return null;
    }

    @Override
    public Account deleteAccount(int id) {
        try {
            String sql = "DELETE FROM accounts WHERE acctId = ? RETURNING *";
            PreparedStatement ps = conn.prepareStatement(sql);

            ps.setInt(1, id);
            ResultSet rs = ps.executeQuery();

            if (rs.next()) {
                return buildAccount(rs);
            }

        } catch (SQLException e) {
            e.printStackTrace();
        }

        return null;
    }


    //Helper Method
    private Account buildAccount(ResultSet rs) throws SQLException {
        Account a = new Account();
        a.setAcctId(rs.getInt("acctId"));
        a.setAcct(rs.getString("accounttype"));
        a.setUserId(rs.getInt("userId"));
        a.setAcctBal(rs.getDouble("accountBalance"));

        return a;
    }
}